src/maxwell-web-view.c

```diff
@@ -0,0 +1,898 @@
+
+  if (JSValueIsArray (context, value))
+    {
+      JSObjectRef array = JSValueToObject (context, value, NULL);
+      JSValueRef val;
+      gint i = 0;
+
+      while ((val = JSObjectGetPropertyAtIndex (context, array, i, NULL)) &&
+             JSValueIsObject (context, val))
+        {
+          JSObjectRef obj = JSValueToObject (context, val, NULL);
+          gchar *child_id = _js_object_get_string (context, obj, "id");
+          ChildData *data = get_child_data_by_id (priv, child_id);
+
+          if (data)
+            {
+              data->alloc.x = _js_object_get_number (context, obj, "x");
+              data->alloc.y = _js_object_get_number (context, obj, "y");
+            }
+
+          g_free (child_id);
+          i++;
+        }
+    }
+  else
+    {
+      g_warning ("Error running javascript: unexpected return value");
+    }
+}
+
+static gboolean
+child_allocate (ChildData *data)
+{
+  GtkRequisition natural;
+  GtkAllocation alloc;
+
+  gtk_widget_get_preferred_size (data->child, NULL, &natural);
+
+  if (natural.width == data->alloc.width && natural.height == data->alloc.height)
+    return FALSE;
+
+  alloc.x = 0;
+  alloc.y = 0;
+  alloc.height = natural.height;
+  alloc.width = natural.width;
+
+  /* Update canvas alloc size */
+  data->alloc.width = natural.width;
+  data->alloc.height = natural.height;
+
+  gtk_widget_size_allocate (data->child, &alloc);
+
+  if (data->offscreen)
+    gdk_window_resize (data->offscreen, natural.width, natural.height);
+
+  return TRUE;
+}
+
+static void
+child_update_visibility (MaxwellWebView *webview, GtkWidget *child)
+{
+  MaxwellWebViewPrivate *priv = MAXWELL_WEB_VIEW_PRIVATE (webview);
+  ChildData *data = get_child_data_by_child (priv, child);
+
+  if (priv->script_loaded && data && data->id)
+    _js_run (WEBKIT_WEB_VIEW (webview),
+             "maxwell.child_set_visible ('%s', %s);",
+             data->id,
+             gtk_widget_get_visible (child) ? "true" : "false");
+}
+
+static void
+handle_script_message_children_allocate (WebKitUserContentManager *manager,
+                                         WebKitJavascriptResult   *result,
+                                         MaxwellWebView           *webview)
+{
+  MaxwellWebViewPrivate *priv = MAXWELL_WEB_VIEW_PRIVATE (webview);
+  JSGlobalContextRef context = webkit_javascript_result_get_global_context (result);
+  JSValueRef value = webkit_javascript_result_get_value (result);
+
+  if (JSValueIsArray (context, value))
+    {
+      JSObjectRef array = JSValueToObject (context, value, NULL);
+      GString *script = g_string_new ("");
+      JSValueRef val;
+      gint i = 0;
+
+      while ((val = JSObjectGetPropertyAtIndex (context, array, i, NULL)) &&
+             JSValueIsString (context, val))
+        {
+          gchar *id = _js_get_string (context, val);
+          ChildData *data = get_child_data_by_id (priv, id);
+
+          if (data)
+            {
+              child_allocate (data);
```

Comment:
Do we not end up calling `gtk_widget_size_allocate()` outside of the allocation cycle as a consequence of this?

Reply:
Yes, I will see if I can simplify this, adding a comment for now